Discrepancy with combination of (Shared Thread pool and Method Interceptor)

I'm trying to setup a simple/proof of concept project to see the new features regarding the maximization of parallelization.
Im encountering the following error after what seems to be the "first" batch of data-provider cases.

```
[ERROR] tests.DemoTest.testStrings -- Time elapsed: 0.070 s <<< FAILURE!
java.lang.ClassCastException: class java.util.concurrent.CompletableFuture$AsyncSupply cannot be cast to class java.lang.Comparable (java.util.concurrent.CompletableFuture$AsyncSupply and java.lang.Comparable are in module java.base of loader 'bootstrap')
        at java.base/java.util.concurrent.PriorityBlockingQueue.siftUpComparable(PriorityBlockingQueue.java:349)
        at java.base/java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:475)
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1368)
        at java.base/java.util.concurrent.CompletableFuture.asyncSupplyStage(CompletableFuture.java:1782)
        at java.base/java.util.concurrent.CompletableFuture.supplyAsync(CompletableFuture.java:2005)
        at org.testng.internal.invokers.MethodRunner.runInParallel(MethodRunner.java:143)
        at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:950)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:201)
        at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:148)
        at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at org.testng.internal.thread.graph.TestNGFutureTask.run(TestNGFutureTask.java:22)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```


More details:
The test class:

```
package tests;

import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.ArrayList;
import java.util.List;

public class DemoTest {

    @DataProvider(parallel = true)
    public static Object[] parallelDpStrings() {
        List<String> l = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            l.add(STR."string \{i}");
        }
        return l.toArray();
    }

    @Test(dataProvider = "parallelDpStrings")
    public void testStrings(String s) throws InterruptedException {
        print(s);
        Thread.sleep(3000);
    }

    public static void print(String s) {
        System.out.println(Thread.currentThread().threadId() + "  " + Thread.currentThread().getStackTrace()[2].getMethodName() + " " + s);
    }
}
```

testng.xml file:
```
<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.1.dtd">
<suite name="testng.xml"
       parallel="instances"
       use-global-thread-pool="true"
       share-thread-pool-for-data-providers="true"
>

    <test name="tng">
        <packages>
            <package name="tests.*"/>
        </packages>
    </test>
</suite>
```

maven command:

`mvn clean test -DthreadCount=10`

more console output:
```
[INFO] --- surefire:3.2.1:test (default-test) @ tng-gtc ---
[INFO] Using auto detected provider org.apache.maven.surefire.testng.TestNGProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running TestSuite
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
42  testStrings string 8
40  testStrings string 6
35  testStrings string 1
41  testStrings string 7
38  testStrings string 4
36  testStrings string 2
39  testStrings string 5
34  testStrings string 0
37  testStrings string 3
[ERROR] Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.457 s <<< FAILURE! -- in TestSuite
[ERROR] tests.DemoTest.testStrings -- Time elapsed: 0.070 s <<< FAILURE!
java.lang.ClassCastException: class java.util.concurrent.CompletableFuture$AsyncSupply cannot be cast to class java.lang.Comparable (java.util.concurrent.CompletableFuture$AsyncSupply and java.lang.Comparable are in module java.base of loader 'bootstrap')
        at java.base/java.util.concurrent.PriorityBlockingQueue.siftUpComparable(PriorityBlockingQueue.java:349)
        at java.base/java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:475)
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1368)
        at java.base/java.util.concurrent.CompletableFuture.asyncSupplyStage(CompletableFuture.java:1782)
        at java.base/java.util.concurrent.CompletableFuture.supplyAsync(CompletableFuture.java:2005)
        at org.testng.internal.invokers.MethodRunner.runInParallel(MethodRunner.java:143)
        at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:950)
        at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:201)
        at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:148)
        at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at org.testng.internal.thread.graph.TestNGFutureTask.run(TestNGFutureTask.java:22)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DemoTest.testStrings » ClassCast class java.util.concurrent.CompletableFuture$AsyncSupply cannot be cast to class java.lang.Comparable (java.util.concurrent.CompletableFuture$AsyncSupply and java.lang.Comparable are in module java.base of loader 'bootstrap')
[INFO] 
[ERROR] Tests run: 10, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.317 s
[INFO] Finished at: 2024-02-29T15:59:43+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.1:test (default-test) on project tng-gtc: There are test failures.
```


Did i miss something?

_Originally posted by @dmirica-gpsw in https://github.com/testng-team/testng/discussions/3080_

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,6 @@
 Current (7.10.0)
+
+Fixed: GITHUB-3081: Discrepancy with combination of (Shared Thread pool and Method Interceptor) (Krishnan Mahadevan)
 Fixed: GITHUB-2381: Controlling the inclusion of the listener at runtime (Krishnan Mahadevan)
 Fixed: GITHUB-3082: IInvokedMethodListener Iinvoked method does not return correct instance during @BeforeMethod, @AfterMethod and @AfterClass (Krishnan Mahadevan)
 Fixed: GITHUB-3084: Document project's PGP artifact signing keys (Krishnan Mahadevan)

--- a/testng-core/src/main/java/org/testng/internal/invokers/MethodRunner.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/MethodRunner.java
@@ -1,7 +1,6 @@
 package org.testng.internal.invokers;
 
 import static java.util.concurrent.CompletableFuture.allOf;
-import static java.util.concurrent.CompletableFuture.supplyAsync;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -23,6 +22,7 @@ import org.testng.internal.ObjectBag;
 import org.testng.internal.Parameters;
 import org.testng.internal.invokers.ITestInvoker.FailureContext;
 import org.testng.internal.invokers.TestMethodArguments.Builder;
+import org.testng.internal.thread.Async;
 import org.testng.internal.thread.TestNGThreadFactory;
 import org.testng.xml.XmlSuite;
 
@@ -140,7 +140,7 @@ public class MethodRunner implements IMethodRunner {
               invocationCount.get(),
               failure.count.get(),
               testInvoker.getNotifier());
-      all.add(supplyAsync(w::call, service));
+      all.add(Async.run(w, service));
       // testng387: increment the param index in the bag.
       parametersIndex += 1;
     }

--- a/testng-core/src/main/java/org/testng/internal/invokers/TestMethodWithDataProviderMethodWorker.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/TestMethodWithDataProviderMethodWorker.java
@@ -14,7 +14,8 @@ import org.testng.internal.TestResult;
 import org.testng.internal.invokers.TestMethodArguments.Builder;
 import org.testng.xml.XmlSuite;
 
-public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITestResult>> {
+public class TestMethodWithDataProviderMethodWorker
+    implements Callable<List<ITestResult>>, Comparable<TestMethodWithDataProviderMethodWorker> {
 
   private final ITestNGMethod m_testMethod;
   private final Object[] m_parameterValues;
@@ -151,5 +152,10 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
 
   public int getInvocationCount() {
     return m_invocationCount;
+  }
+
+  @Override
+  public int compareTo(TestMethodWithDataProviderMethodWorker o) {
+    return Integer.compare(this.m_testMethod.getPriority(), o.m_testMethod.getPriority());
   }
 }

--- a/testng-core/src/main/java/org/testng/internal/thread/Async.java
+++ b/testng-core/src/main/java/org/testng/internal/thread/Async.java
@@ -1,0 +1,46 @@
+package org.testng.internal.thread;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import org.testng.ITestResult;
+import org.testng.internal.invokers.TestMethodWithDataProviderMethodWorker;
+
+public final class Async {
+
+  private Async() {
+    // Defeat instantiation
+  }
+
+  public static CompletableFuture<List<ITestResult>> run(
+      TestMethodWithDataProviderMethodWorker worker, ExecutorService service) {
+    AsyncTask asyncTask = new AsyncTask(worker);
+    service.execute(asyncTask);
+    return asyncTask.result;
+  }
+
+  private static class AsyncTask implements Runnable, Comparable<AsyncTask> {
+    private final CompletableFuture<List<ITestResult>> result = new CompletableFuture<>();
+    private final TestMethodWithDataProviderMethodWorker worker;
+
+    public AsyncTask(TestMethodWithDataProviderMethodWorker worker) {
+      this.worker = worker;
+    }
+
+    @Override
+    public void run() {
+      try {
+        if (!result.isDone()) {
+          result.complete(worker.call());
+        }
+      } catch (Throwable t) {
+        result.completeExceptionally(t);
+      }
+    }
+
+    @Override
+    public int compareTo(AsyncTask o) {
+      return worker.compareTo(o.worker);
+    }
+  }
+}

--- a/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
+++ b/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
@@ -59,6 +59,8 @@ import test.dataprovider.issue3041.SampleTestCase;
 import test.dataprovider.issue3045.DataProviderListener;
 import test.dataprovider.issue3045.DataProviderTestClassSample;
 import test.dataprovider.issue3045.DataProviderWithoutListenerTestClassSample;
+import test.dataprovider.issue3081.NoOpMethodInterceptor;
+import test.dataprovider.issue3081.TestClassWithPrioritiesSample;
 
 public class DataProviderTest extends SimpleBaseTest {
 
@@ -649,6 +651,45 @@ public class DataProviderTest extends SimpleBaseTest {
     assertThat(listener.getThreadIds())
         .withFailMessage("Ensuring that the thread ids are correct")
         .hasSize(pair.second());
+  }
+
+  @Test(description = "GITHUB-3081")
+  public void ensureNoExceptionsWhenRunningInSharedThreadPoolsWithMethodInterceptorsNoPriorities() {
+    TestNG testng = create(test.dataprovider.issue3081.TestClassSample.class);
+    test.dataprovider.issue3081.TestClassSample.clear();
+    testng.shouldUseGlobalThreadPool(true);
+    testng.addListener(new NoOpMethodInterceptor());
+    testng.setThreadCount(10);
+    testng.setParallel(XmlSuite.ParallelMode.METHODS);
+    testng.shareThreadPoolForDataProviders(true);
+    testng.setVerbose(2);
+    testng.run();
+    assertThat(testng.getStatus()).isEqualTo(0);
+    assertThat(test.dataprovider.issue3081.TestClassSample.getLogs())
+        .withFailMessage(
+            "There should have been 9 threads ONLY used by the data driven test "
+                + "because one thread would be the main thread on which TestNG would be running")
+        .hasSize(9);
+  }
+
+  @Test(description = "GITHUB-3081")
+  public void
+      ensureNoExceptionsWhenRunningInSharedThreadPoolsWithMethodInterceptorsWithPriorities() {
+    TestNG testng = create(TestClassWithPrioritiesSample.class);
+    TestClassWithPrioritiesSample.clear();
+    testng.shouldUseGlobalThreadPool(true);
+    testng.addListener(new NoOpMethodInterceptor());
+    testng.setParallel(XmlSuite.ParallelMode.METHODS);
+    testng.shareThreadPoolForDataProviders(true);
+    testng.setThreadCount(10);
+    testng.setVerbose(2);
+    testng.run();
+    assertThat(testng.getStatus()).isEqualTo(0);
+    assertThat(TestClassWithPrioritiesSample.getLogs())
+        .withFailMessage(
+            "There should have been 9 threads ONLY used by the data driven test "
+                + "because one thread would be the main thread on which TestNG would be running")
+        .hasSize(9);
   }
 
   @DataProvider

--- a/testng-core/src/test/java/test/dataprovider/issue3081/NoOpMethodInterceptor.java
+++ b/testng-core/src/test/java/test/dataprovider/issue3081/NoOpMethodInterceptor.java
@@ -1,0 +1,13 @@
+package test.dataprovider.issue3081;
+
+import java.util.List;
+import org.testng.IMethodInstance;
+import org.testng.IMethodInterceptor;
+import org.testng.ITestContext;
+
+public class NoOpMethodInterceptor implements IMethodInterceptor {
+  @Override
+  public List<IMethodInstance> intercept(List<IMethodInstance> methods, ITestContext context) {
+    return methods;
+  }
+}

--- a/testng-core/src/test/java/test/dataprovider/issue3081/TestClassSample.java
+++ b/testng-core/src/test/java/test/dataprovider/issue3081/TestClassSample.java
@@ -1,0 +1,39 @@
+package test.dataprovider.issue3081;
+
+import java.security.SecureRandom;
+import java.util.Collections;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class TestClassSample {
+  private static final Set<Long> logs = ConcurrentHashMap.newKeySet();
+  private static final Random random = new SecureRandom();
+
+  public static Set<Long> getLogs() {
+    return Collections.unmodifiableSet(logs);
+  }
+
+  public static void clear() {
+    logs.clear();
+  }
+
+  @DataProvider(parallel = true)
+  public static Object[] parallelDpStrings() {
+    return IntStream.rangeClosed(0, 99).mapToObj(it -> "string " + it).toArray(String[]::new);
+  }
+
+  @Test(dataProvider = "parallelDpStrings")
+  public void testStrings(String ignored) throws InterruptedException {
+    print();
+    TimeUnit.MILLISECONDS.sleep(random.nextInt(500));
+  }
+
+  private static void print() {
+    logs.add(Thread.currentThread().getId());
+  }
+}

--- a/testng-core/src/test/java/test/dataprovider/issue3081/TestClassWithPrioritiesSample.java
+++ b/testng-core/src/test/java/test/dataprovider/issue3081/TestClassWithPrioritiesSample.java
@@ -1,0 +1,42 @@
+package test.dataprovider.issue3081;
+
+import java.security.SecureRandom;
+import java.util.Collections;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class TestClassWithPrioritiesSample {
+  private static final Set<Long> logs = ConcurrentHashMap.newKeySet();
+  private static final Random random = new SecureRandom();
+
+  public static Set<Long> getLogs() {
+    return Collections.unmodifiableSet(logs);
+  }
+
+  public static void clear() {
+    logs.clear();
+  }
+
+  @DataProvider(parallel = true)
+  public static Object[] parallelDpStrings() {
+    return IntStream.rangeClosed(0, 99).mapToObj(it -> "string " + it).toArray(String[]::new);
+  }
+
+  @Test(dataProvider = "parallelDpStrings", priority = 1)
+  public void testStrings(String ignored) throws InterruptedException {
+    print();
+    TimeUnit.MILLISECONDS.sleep(random.nextInt(500));
+  }
+
+  @Test(priority = 2)
+  public void anotherTest() {}
+
+  private static void print() {
+    logs.add(Thread.currentThread().getId());
+  }
+}

